### PR TITLE
Make Maestro trigger the release pipeline associated with a channel

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/1543

### CODE REVIEW

File: src/Maestro/Maestro.Web/Startup.cs [hunk 1]
```diff
@@ -45,14 +49,89 @@ static Startup()
         {
             Triggers<BuildChannel>.Inserted += entry =>
             {
-                BuildChannel entity = entry.Entity;
                 DbContext context = entry.Context;
+                BuildChannel entity = entry.Entity;
+
+                StartAssociatedReleasePipelinesAsync(entry).GetAwaiter().GetResult();
```
Comment (JohnTortugo) [inline src/Maestro/Maestro.Web/Startup.cs]: We'll make this method a new service. The service will be called in a similar way as we do for DependencyUpdater below.  Tracked by this issue: https://github.com/dotnet/arcade/issues/1994

File: src/Maestro/Maestro.Web/Startup.cs [hunk 2]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+            // If something use the old API version we won't have this information available.
+            // This will also be the case if something adds an existing build (created using 
+            // the old API version) to a channel
+            if (barBuildInfo.AzureDevOpsBuildId == null)
+            {
+                logger.LogWarning($"barBuildInfo.AzureDevOpsBuildId is null for BAR Build.Id {barBuildInfo.Id}.");
+                return;
+            }
+
+            if (channel.ChannelReleasePipelines?.Any() != true)
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Startup.cs]: `if (!channel.ChannelReleasePipelines?.Any())`

Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.cs]: I'm not sure but will "!" convert "null" to false?

File: src/Maestro/Maestro.Web/Startup.cs [hunk 3]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+
+            foreach (var pipeline in channel.ChannelReleasePipelines)
+            {
+                try
+                {
+                    var account = pipeline.ReleasePipeline.Organization;
+                    var project = pipeline.ReleasePipeline.Project;
+                    var pipelineId = pipeline.ReleasePipeline.PipelineIdentifier;
+
+                    var pipeDef = await azdoClient.GetReleaseDefinitionAsync(account, project, pipelineId);
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Startup.cs]: Are we going to run into concurrency issues if multiple runs of the release pipeline happen at the same time? Does this segment need to lock?

Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.cs]: That's possible, although seems the chance of it happening are very low. @riarenas is working on an improvement on how we trigger the pipeline that will fix this problem as well: https://github.com/dotnet/arcade/issues/1994

File: src/Maestro/Maestro.Web/Startup.cs (hunk 2 above)
Comment (mmitche) [reply src/Maestro/Maestro.Web/Startup.cs]: It might? Either it's probably fine.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: GetAwaiter().GetResult() is a recipe for deadlocking the service.  Is this really something that can't be synchronous?

**Comment by JohnTortugo** [comment]
I tested this flow by connecting my local MaestroApplication to BAR INT and was able to trigger the [release pipeline here](https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=1346) for some test Arcade builds.

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs [hunk 4]
```diff
@@ -884,6 +887,55 @@ public async void StartNewReleaseAsync(string accountName, string projectName, A
...
+        /// <summary>
+        /// Block until the release with ID informed finishes execution.
+        /// Checks the status of the release every 15 seconds.
+        /// </summary>
+        /// <param name="accountName"></param>
+        /// <param name="projectName"></param>
+        /// <param name="releaseId"></param>
+        public void WaitUntilCompleted(string accountName, string projectName, int releaseId)
+        {
+            while (true)
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: You need some level of try/catch error handling here so that intermittent issues (e.g. random 500 from AzDO) don't cause havoc.

I'd also put in a max timeout.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 2 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: I'm pretty sure the ! operator will lift, leaving you with still nullable thing.  but it might be clearer just to separate out the null check, so the mind bending of what happens if it's null, or empty, and what does != true do for null.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: Especially given there are web calls involved there, I don't think this is appropriate. It should quickly be shoved into a queue somewhere, and a reliable (and async) queue processor can handle starting the job.

Comment (alexperovich) [reply src/Maestro/Maestro.Web/Startup.cs]: That is what the BackgroundQueue thing is for. Its a "I need to send off this async work and not block my thread" his should use that.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: We appear to already have a "BackgroundQueue" thing a few lines down. Can we use that?

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs (hunk 4 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: Yeah, I was looking for exactly that kind of feedback here!

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (mmitche) [reply src/Maestro/Maestro.Web/Startup.cs]: @ChadNedzlek How did it deadlock the Helix API? Based on my understanding, this is essentially just turning the async call into sync.  Anyways, the API itself is async since it eventually devolves into REST API calls.  So at some level it will need to be GetAwaiter().GetResult(), or turned into async here.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 5]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+
+                    pipeDef = await azdoClient.RemoveAllArtifactSourcesAsync(account, project, pipeDef);
+
+                    pipeDef = await azdoClient.AddArtifactSourceAsync(account, project, pipeDef, azdoBuild);
+
+                    var releaseId = await azdoClient.StartNewReleaseAsync(account, project, pipeDef, barBuildInfo.Id);
+
+                    azdoClient.WaitUntilCompleted(account, project, releaseId);
+                }
+                catch (HttpRequestException e)
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: Are these much better than the generic catch latter? It's just slightly changing the message (and removing the identifier) I'd probably just leave the general handler.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.cs]: The issue that I pointed in the first message here is to do exactly that. Use BackgroundQueue as a way to execute `StartAssociatedReleasePipelinesAsync`.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 6]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+                logger.LogWarning($"Channel {channel.Id}, which build with BAR ID {barBuildInfo.Id} is attached to, doesn't have an associated publishing pipeline.");
+                return;
+            }
+
+            var azdoTokenProvider = context.GetService<IAzureDevOpsTokenProvider>();
+            var accessToken = azdoTokenProvider.GetTokenForAccount(barBuildInfo.AzureDevOpsAccount).GetAwaiter().GetResult();
+
+            var azdoClient = new AzureDevOpsClient(accessToken, logger, null);
+
+            var azdoBuild = await azdoClient.GetBuildAsync(
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: There is no retry where and Azure DevOps is notoriously flaky.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 7]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+                barBuildInfo.AzureDevOpsAccount,
+                barBuildInfo.AzureDevOpsProject,
+                barBuildInfo.AzureDevOpsBuildId.Value);
+
+            foreach (var pipeline in channel.ChannelReleasePipelines)
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: Most of these "var" need to be made explicit according to our coding style.

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs [hunk 8]
```diff
@@ -884,6 +887,55 @@ public async void StartNewReleaseAsync(string accountName, string projectName, A
                 body,
                 versionOverride: "5.0-preview.3",
                 baseAddressSubpath: "vsrm.");
+
+            return content.GetValue("id").ToObject<int>();
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: I think this can be content.Value<int>("id")

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Startup.cs]: The problem with .GetAwaiter().GetResult() is that it blocks the calling thread until the Task that was launched is done. That causes a deadlock if the calling thread is needed to complete the task.

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs [hunk 9]
```diff
@@ -884,6 +887,55 @@ public async void StartNewReleaseAsync(string accountName, string projectName, A
...
+        /// Checks the status of the release every 15 seconds.
+        /// </summary>
+        /// <param name="accountName"></param>
+        /// <param name="projectName"></param>
+        /// <param name="releaseId"></param>
+        public void WaitUntilCompleted(string accountName, string projectName, int releaseId)
+        {
+            while (true)
+            {
+                AzureDevOpsRelease rel = GetReleaseAsync(accountName, projectName, releaseId).GetAwaiter().GetResult();
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: This is another deadlock problem.  Who is calling this that needs to spend a thread in this busy loop, and why aren't they async in the first place?

Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: The only caller of this is, itself, async. This whole method should be async, and they should wait for it... You have sync-over-async-over-sync-over-async.  This needs to get cleaned up so you stop switching contexts.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 10]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+
+                StartAssociatedReleasePipelinesAsync(entry).GetAwaiter().GetResult();
+
                 var queue = context.GetService<BackgroundQueue>();
                 var dependencyUpdater = context.GetService<IDependencyUpdater>();
                 queue.Post(() => dependencyUpdater.StartUpdateDependenciesAsync(entity.BuildId, entity.ChannelId));
             };
         }
 
+        private static async Task StartAssociatedReleasePipelinesAsync(IInsertedEntry<BuildChannel, DbContext> entry)
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: This method, even though called "Start" seems to actually wait. It's not named right. Its really "ExecuteAssociatedReleasePipelinesAsync".

I'm not sure we want a bunch of loops all running at once here, all waiting 15 seconds for it to finish.  One reboot of this service and we are screwed, we'll drop all of them on the floor.

This needs to be done reliably.  We need a store of "all releases that are pending" and we check them all and update the store and do whatever work is necessary.

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs [hunk 11]
```diff
@@ -884,6 +887,55 @@ public async void StartNewReleaseAsync(string accountName, string projectName, A
...
+
+                if (rel.Environments[0].Status != "notStarted"
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: I'd probably do a switch/case here to make it less noisy... I'd probably also ToLowerInvariant() the value, because these are... suspiciously cased.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 12]
```diff
@@ -45,14 +49,89 @@ static Startup()
...
+        {
+            BuildAssetRegistryContext context = entry.Context as BuildAssetRegistryContext;
+            ILogger<BuildChannel> logger = context.GetService<ILogger<BuildChannel>>();
+            BuildChannel entity = entry.Entity;
+            Build barBuildInfo = entity.Build;
+            Channel channel = context.Channels
+                .Where(ch => ch.Id == entity.ChannelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .First();
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: FirstAsync + await

File: src/Maestro/Maestro.Web/Startup.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: Helpful article by our local async expert, Stephen Toub: https://blogs.msdn.microsoft.com/pfxteam/2012/04/13/should-i-expose-synchronous-wrappers-for-asynchronous-methods/

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: oh! An even better one: https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

TL;DR: Don't ever call GetAwaiter().GetResult().  If you absolutely, super, definitely have to, you need to make sure literally every await in the entire closure of your entire program uses "ConfigureAwait(false)" (which is basically impossible, and you'll probably spend a huge amount of time tracking down random deadlocks as long as the code exists)

File: src/Maestro/Maestro.Web/Startup.cs (hunk 10 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Startup.cs]: This logic needs to all be in another StatefulService so it can store information about release pipelines it has started and check when they are finished.

**Comment by riarenas** [comment]
I believe all the required changes are here now. Please let me know if you'd rather I close this PR and open a new one if this is too confusing now.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 13]
```diff
@@ -21,20 +21,24 @@
 using Microsoft.AspNetCore.Rewrite;
 using Microsoft.Azure.KeyVault;
 using Microsoft.Azure.KeyVault.Models;
+using Microsoft.DotNet.DarcLib;
 using Microsoft.DotNet.ServiceFabric.ServiceHost;
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Startup.cs]: What is this using for?

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.Extensions.Logging;
+using Microsoft.ServiceFabric.Data;
+using Microsoft.ServiceFabric.Data.Collections;
+using Newtonsoft.Json;
+
+namespace ReleasePipelineRunner
+{
+    [DataContract]
+    public sealed class ReleasePipelineRunnerItem
```
Comment (alexperovich) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: NIT: move these types into matching files.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 13 above)
Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: AzureDevopsClient

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 14 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Sure, can do that. I only did it here because DependencyUpdater also has these kind of data items specified in the class.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 15]
```diff
@@ -43,16 +47,40 @@ public partial class Startup
...
         static Startup()
         {
-            Triggers<BuildChannel>.Inserted += entry =>
+            Triggers<BuildChannel>.Inserting += entry =>
             {
+                BuildAssetRegistryContext context = entry.Context as BuildAssetRegistryContext;
                 BuildChannel entity = entry.Entity;
+                if (ChannelHasAssociatedReleasePipeline(entity.ChannelId, context))
+                {
+                    entry.Cancel = true;
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: >entry.Cancel = true; [](start = 20, length = 20)

Cancel? Does this stop it from inserting into the DB??? What does that do to the caller?

File: src/Maestro/Maestro.Web/Startup.cs [hunk 16]
```diff
@@ -43,16 +47,40 @@ public partial class Startup
...
             };
         }
 
+        private static bool ChannelHasAssociatedReleasePipeline(int channelId, BuildAssetRegistryContext context)
+        {
+            Channel channel = context.Channels
+                .Where(ch => ch.Id == channelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .FirstOrDefaultAsync().GetAwaiter().GetResult();
```
Comment (ChadNedzlek) [inline src/Maestro/Maestro.Web/Startup.cs]: >.FirstOrDefaultAsync().GetAwaiter().GetResult(); [](start = 16, length = 48)

This is deadlocks again... Aren't there non-async versions of FirstOrDefault?

Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: Sure, I can use the non-async

File: src/Maestro/Maestro.Web/Startup.cs (hunk 15 above)
Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: Yes, it stops it from being set in the DB. The caller still sees a 201.

Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs]: This is... super hacky.  We shouldn't be using a DB insert to trigger a build and not actually insert into the database. It's extremely unexpected for something like `await _context.BuildChannels.AddAsync(buildChannel);` doesn't actually insert into the database. And it means the web call will say "I added it to the channel", but then if you query that build, it won't be in that channel.

Really the callers that are calling "add to channel" shouldn't be doing that. We should fix the callers so that they call "DoAppropriateReleaseStuffAsync" instead, and that does the right thing.

Having a trigger you can't see in the code that cancels the insert and lies to a web request that schedules a magic release that implicitly will fix... whatever... is way too much implicit magic and is a maintenance nightmare.  It took Ricardo and I almost 15 minutes just to find where we _think_ this code was being triggered from, and we're only like 50% sure that's actually the right place.  Triggers are *_BAD_*.

---
In reply to: [260462310](https://github.com/dotnet/arcade-services/pull/125#discussion_r260462310) [](ancestors = 260462310)

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+                                item.ChannelId))
+                            {
+                                await RunAssociatedReleasePipelinesAsync(item.BuildId, item.ChannelId, cancellationToken);
+                            }
+                        }
+
+                        await tx.CommitAsync();
+                    }
+
+                    await Task.Delay(1000, cancellationToken);
```
Comment (ChadNedzlek) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: > await Task.Delay(1000, cancellationToken); [](start = 19, length = 43)

We probably only want to delay if we didn't find work to do. No reason to have the thing sitting around when there is a pile of stuff in the queue. :-)

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+                        await runningPipelines.AddOrUpdateAsync(tx, buildId, releaseList, (key, oldValue) => releaseList);
+                    }
+                    else
+                    {
+                        await runningPipelines.AddAsync(tx, buildId, releaseList);
+                    }
+                    await tx.CommitAsync();
+                }
+            }
+            await ProcessFinishedReleasesAsync(cancellationToken);
```
Comment (alexperovich) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: This isn't necessary. Just let the schedule get it.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+            {
+                HashSet<BuildChannel> buildChannelsToAdd = new HashSet<BuildChannel>(new BuildChannelComparer());
+                using (ITransaction tx = StateManager.CreateTransaction())
+                {
+                    var runningPipelinesEnumerable = await runningPipelines.CreateEnumerableAsync(tx, EnumerationMode.Unordered);
+                    var asyncEnumerator = runningPipelinesEnumerable.GetAsyncEnumerator();
+                    while (await asyncEnumerator.MoveNextAsync(cancellationToken))
+                    {
+                        int buildId = asyncEnumerator.Current.Key;
+                        ConditionalValue<IList<ReleasePipelineStatusItem>> maybeReleaseStatuses = await runningPipelines.TryGetValueAsync(tx, buildId);
```
Comment (alexperovich) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: This isn't needed. You have the value already in `asyncEnumerator.Current.Value`

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+                    string organization = pipeline.ReleasePipeline.Organization;
+                    string project = pipeline.ReleasePipeline.Project;
+                    int pipelineId = pipeline.ReleasePipeline.PipelineIdentifier;
+
+                    AzureDevOpsReleaseDefinition pipeDef = await azdoClient.GetReleaseDefinitionAsync(organization, project, pipelineId);
+                    pipeDef = await azdoClient.RemoveAllArtifactSourcesAsync(organization, project, pipeDef);
+
+                    pipeDef = await azdoClient.AddArtifactSourceAsync(organization, project, pipeDef, azdoBuild);
+
+                    int releaseId = await azdoClient.StartNewReleaseAsync(organization, project, pipeDef, build.Id);
```
Comment (ChadNedzlek) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Are these command safe to be run in paralle? I can't really tell what they are doing, but I think they are messing around with the artifact sources of a pipeline definiton, which might mean if 2 things are looking at the same definition, they'll be adding and removing stuff, causing weird conflicts.

Do I misunderstand this? The ReleasePipelineRunner is set to 3 instances, so it might be fighting with itself here. We will probably need some sort of "I'm locking this pipeline definition, no one is allowed to touch it" type thing.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+        /// Check for finished releases and add the corresponding build channel association to the DB.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        // Check every 10 minutes for any finished release Pipelines
+        [CronSchedule("0 0/10 * 1/1 * ? *", TimeZones.UTC)]
+        public async Task ProcessFinishedReleasesAsync(CancellationToken cancellationToken)
+        {
+            var runningPipelines =
+                        await StateManager.GetOrAddAsync<IReliableDictionary<int, IList<ReleasePipelineStatusItem>>>(RunningPipelineDictionaryName);
```
Comment (ChadNedzlek) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: This indentation is weird: why so far indented?  It didn't make the line much smaller either.  Unfortunately, most of it is the huge generic, so I don't have a good line breaking plan that looks good. :-)

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+                        {
+                            bool allFinished = true;
+                            IList<ReleasePipelineStatusItem> releaseStatuses = maybeReleaseStatuses.Value;
+
+                            foreach (ReleasePipelineStatusItem releaseStatus in releaseStatuses)
+                            {
+                                int releaseId = releaseStatus.ReleaseId;
+                                AzureDevOpsClient azdoClient = await GetAzureDevOpsClientForAccount(releaseStatus.PipelineOrganization);
+                                AzureDevOpsRelease release = await azdoClient.GetReleaseAsync(releaseStatus.PipelineOrganization, releaseStatus.PipelineProject, releaseStatus.ReleaseId);
+                                string currentStatus = release.Environments[0].Status;
```
Comment (ChadNedzlek) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: >[0] [](start = 75, length = 3)

Should we check all the environments? Or log something if there is not exactly one?

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+                                if (InProgressStatuses.Contains(currentStatus.ToLowerInvariant()))
```
Comment (ChadNedzlek) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: >currentStatus.ToLowerInvariant() [](start = 64, length = 32)

If you made the InProgressStatuses use OrdinalIgnoreCase, you wouldn't have to tolower everything. :-)

File: src/Microsoft.DotNet.ServiceFabric.ServiceHost/LoggedServiceRemoting/LoggingServiceProxyInterceptor.cs [hunk 24]
```diff
@@ -97,7 +97,7 @@ protected override T Intercept<T>(IInvocation invocation, Func<T> call)
             {
                 try
                 {
-                    Activity.Current.AddBaggage("CallingServiceName", Context.ServiceName.ToString());
+                    Activity.Current.AddBaggage("CallingServiceName", $"\"{Context.ServiceName.ToString()}\"");
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.ServiceFabric.ServiceHost/LoggedServiceRemoting/LoggingServiceProxyInterceptor.cs]: >ToString [](start = 95, length = 8)

NIT: Since it's being interpolated now, I think the "ToString()" is redundant.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+                                        BuildId = buildId,
+                                        ChannelId = releaseStatus.ChannelId
+                                    });
+                                }
+                            }
+
+                            // Stop tracking releases for builds where every release finished.
+                            if (allFinished)
+                            {
+                                await runningPipelines.TryRemoveAsync(tx, asyncEnumerator.Current.Key);
```
Comment (riarenas) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Changing the implementation of this due to a possible case where this list becomes infinitely long that @alexperovich pointed out.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 22 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: I'll check the others. It's the correct thing to do

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 21 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Weird indentation was probably from refactoring something at some point and I just didn't realize.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 23 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: I always forgets that exists, and it's so nice

File: src/Maestro/Maestro.Web/Startup.cs (hunk 15 above)
Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: Chatted offline.

It's a little late to change the design at this point. Definitely something to consider as a future improvement.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 19 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Yep. That was a derp while just trying to learn about the reliable collections. fixed now

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 20 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Discussed offline. Right now they won't run in parallel. Once this goes in I'll create a follow up issue to consider improving this part.

**Comment by riarenas** [comment]
I believe all feedback has been incorporated except
1. Won't change the trigger design.
2. Won't invest in making starting the release safe to run in parallel.

PTAL

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 20 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: RunAsync only runs in the primary instance. This is a valid concern if we ever want to do this in parallel. Once this goes in I'll file an issue so we keep this limitation in mind and find alternatives.

**Comment by riarenas** [comment]
@mmitche @ChadNedzlek any additional concerns?

**Comment by ChadNedzlek** [comment]
Given the current scope, it's acceptable to check this in.

**Comment by mmitche** [comment]
Taking a look.

File: src/Maestro/Maestro.Web/Startup.cs [hunk 26]
```diff
@@ -43,16 +47,40 @@ public partial class Startup
...
             };
         }
 
+        private static bool ChannelHasAssociatedReleasePipeline(int channelId, BuildAssetRegistryContext context)
+        {
+            Channel channel = context.Channels
+                .Where(ch => ch.Id == channelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .FirstOrDefault();
```
Comment (mmitche) [inline src/Maestro/Maestro.Web/Startup.cs]: It might be clearer to just put the condition in the FirstOrDefault:

return context.Channels
                .Where(ch => ch.Id == channelId)
                .Include(ch => ch.ChannelReleasePipelines)
                .ThenInclude(crp => crp.ReleasePipeline)
                .FirstOrDefault(c => c.ChannelReleasePipelines.Count > 0);

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+                    using (ITransaction tx = StateManager.CreateTransaction())
+                    {
+                        ConditionalValue<ReleasePipelineRunnerItem> maybeItem = await queue.TryDequeueAsync(
+                            tx,
+                            cancellationToken);
+                        if (maybeItem.HasValue)
+                        {
+                            ReleasePipelineRunnerItem item = maybeItem.Value;
+                            using (Logger.BeginScope(
+                                "Triggering release pipelines associated with channel {channelId} for build {buildId}.",
```
Comment (mmitche) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: I think you're missing a $ prefix for string interpolation here.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs [hunk 28]
```diff
@@ -0,0 +1,294 @@
...
+
+            Channel channel = await Context.Channels
+                .Where(ch => ch.Id == channelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .FirstOrDefaultAsync();
+
+            if (build == null || channel == null)
+            {
+                Logger.LogInformation($"Could not find the specified build {buildId} or channel {channelId} to run a release pipeline on");
```
Comment (mmitche) [inline src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Should this be an error? At least if build == null, since builds can't be deleted.  Channels can be deleted, but it's a pretty exceptional case.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 27 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: That's a good catch. We have the same bug in DependencyUpdater where I stole this code from. I'll update it there as well.

File: src/Maestro/Maestro.Web/Startup.cs (hunk 26 above)
Comment (riarenas) [reply src/Maestro/Maestro.Web/Startup.cs]: I like that.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 28 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: I'll split it and make it an error if Build is null.

File: src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs (hunk 27 above)
Comment (riarenas) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: Nevermind! synced with @alexperovich this is basically a string.Format

Comment (alexperovich) [reply src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs]: This is actually supposed to not use string interpolation. This is using the Microsoft.Extensions.Logging logger with the string.Format arg style. The names in the {} become property names in the telemetry item.

**Comment by riarenas** [comment]
Performing final round of local testing with the new pats, and then I'll merge.

**Comment by riarenas** [comment]
Fixed a small but and made the task that checks for finished pipelines to check run every 5 minutes instead of 10. 10 was a bit too long for what some of these pipelines do.
